fix: should not tick nested timer in fakeAsync test

JiaLiPassion opened the pull request:
Close #33799

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #33799
```
function waitTwice(callback: () => any): void {
  setTimeout(() =>
    setTimeout(() => callback())
  );
}

describe('fakeAsync', () => {
  it('nested timers', fakeAsync(() => {
    const callback = jasmine.createSpy('callback');
    waitTwice(callback);
    expect(callback).not.toHaveBeenCalled();
    tick();
    expect(callback).toHaveBeenCalled();
    tick();
    expect(callback).toHaveBeenCalled();
  }));
});
```

By default, 
The `callback` will be called after the `first tick`.


```
describe('fakeAsync', () => {
  it('nested timers', fakeAsync(() => {
    const callback = jasmine.createSpy('callback');
    waitTwice(callback);
    expect(callback).not.toHaveBeenCalled();
    tick(0, true);
    expect(callback).not.toHaveBeenCalled();
    tick();
    expect(callback).toHaveBeenCalled();
  }));
});
```
````
## What is the new behavior?
with `tick(0, true)`, the second parameter is `ignoreNestedTimeout`, after the `first tick`, the `nested timeout` will not be invoked.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

mhevery commented:
[presubmit](http://test/OCL:281123086:BASE:281123154:1574107580499:df5f311f)

mhevery commented on packages/zone.js/lib/zone-spec/fake-async-test.ts:
```diff
@@ -83,7 +85,8 @@
     scheduleFunction(
         cb: Function, delay: number, args: any[] = [], isPeriodic: boolean = false,
-        isRequestAnimationFrame: boolean = false, id: number = -1): number {
+        isRequestAnimationFrame: boolean = false, id: number = -1,
```
these function args are getting too long and `id` is inside of the `isRequestAnimationFrame` and `isRequeuePeriodic`. What do you think about refactoring it to have optional named arguments instead?

JiaLiPassion replied:
@mhevery , yeah, you are right, in the updated PR, I refactored the parameters to optional named arguments object.

mhevery commented:
So this change is breaky in `g3` where 30+ targets fail just in blueprint.

mhevery reviewed (commented):
could we add some API docs to the change?

petebacondarwin reviewed (commented):
The last "`fix: format`" commit should be a fixup commit for the previous commit

JiaLiPassion committed:
feat: add an tickOptions parameter with property processNewMacroTasksSynchronously.

This option will control whether to invoke the new macro tasks when ticking.

Close #33799

JiaLiPassion commented:
@mhevery , @petebacondarwin , thank you for the review, I have updated the PR, use new API instead of breaking changes. Please review again, thanks!

petebacondarwin reviewed (approved):
LGTM

mary-poppins commented:
You can preview 461d422 at https://pr33838-461d422.ngbuilds.io/.
You can preview ef9a2c1 at https://pr33838-ef9a2c1.ngbuilds.io/.
You can preview 4ad7ef7 at https://pr33838-4ad7ef7.ngbuilds.io/.
You can preview 69c7eda at https://pr33838-69c7eda.ngbuilds.io/.
You can preview 8d7b564 at https://pr33838-8d7b564.ngbuilds.io/.
You can preview 094d6dc at https://pr33838-094d6dc.ngbuilds.io/.
You can preview 7b39c5b at https://pr33838-7b39c5b.ngbuilds.io/.
You can preview 111ddf4 at https://pr33838-111ddf4.ngbuilds.io/.
You can preview 5f05545 at https://pr33838-5f05545.ngbuilds.io/.
You can preview 6ad4193 at https://pr33838-6ad4193.ngbuilds.io/.

mary-poppins commented:
You can preview d015f32 at https://pr33838-d015f32.ngbuilds.io/.

mhevery commented on packages/core/testing/src/fake_async.ts:
```diff
@@ -62,13 +62,17 @@ export function fakeAsync(fn: Function): (...args: any[]) => any {
  * @publicApi
  */
-export function tick(millis: number = 0): void {
+export function tick(millis: number = 0, ignoreNestedTimeout = false): void {
```
I fear that over time we will have more of these arguments. Could we turn the options into object literal? That way adding new options in the future will not be an issue.?

mhevery replied:
Also I think the name should be `ignoreNestedTimeout` => `processNewMacroTasksSynchronously` (Open to other names)  What do you think. If I understand correctly, this is not specific to `setTimeout` but rather to the fact that if a macro task is scheduled in the current flush cycle, wether or not it should be flushed synchronously.

JiaLiPassion replied:
yes, you are right, it is not only `setTimeout`, the `processNewMacroTasksSynchronously` is a better name, thanks.

mhevery commented on packages/core/testing/src/fake_async.ts:
```diff
@@ -62,13 +62,17 @@ export function fakeAsync(fn: Function): (...args: any[]) => any {
  * {@example core/testing/ts/fake_async.ts region='basic'}
  *
+ * @param millis, the number of millisecond to advance the virtual timer
+ * @param ignoreNestedTimeout, whether to invoke the nested timeout, by default is false,
```
Could you add more text here. Do you mean if the nested timeout will be flushed in the same tick or separately?

JiaLiPassion replied:
Yes, I have added more text and sample codes. Thanks.

mhevery commented:
[presubmit](http://test/OCL:292587640:BASE:292587675:1580500248141:56894f3f)

mhevery reviewed (commented):
Minor suggestions. Running presubmit now.

mary-poppins commented:
You can preview f052067 at https://pr33838-f052067.ngbuilds.io/.

mary-poppins commented:
You can preview 316364a at https://pr33838-316364a.ngbuilds.io/.

mary-poppins commented:
You can preview 4c86886 at https://pr33838-4c86886.ngbuilds.io/.

mary-poppins commented:
You can preview 76b971a at https://pr33838-76b971a.ngbuilds.io/.

mary-poppins commented:
You can preview 246b0bb at https://pr33838-246b0bb.ngbuilds.io/.

mhevery commented:
[presubmit](http://test/OCL:292587640:BASE:294463027:1581444275652:c9c782ac)

mary-poppins commented:
You can preview 284f8b2 at https://pr33838-284f8b2.ngbuilds.io/.

mhevery commented:
[presubmit](http://test/OCL:292587640:BASE:296256842:1582226671611:88bd667)
